### PR TITLE
Q02

### PR DESCRIPTION
**Pull request by hanlonjohn23**
Second Code Challenge.

#### Error Handling
Not sure what is the best way to throw and handle errors, namely in [HttpRequestor](https://github.com/jhanlonhg/scala-challenge/compare/develop...q02#diff-d40692d2f835b6e3cb99fbc6fa0bbbce5ccc56a2e8fc4c44baa2d127be9b8e7fR42) and [Main](https://github.com/jhanlonhg/scala-challenge/compare/develop...q02#diff-a8c936a27247452dcd65a1c15bf1b96626ca6f026fbf345cac89d5fd3c69a2d6R15). Both these errors should be breaking, since they are a response error and a input error, respectively.

#### Bad Patterns
[This solution](https://github.com/jhanlonhg/scala-challenge/compare/develop...q02#diff-a8c936a27247452dcd65a1c15bf1b96626ca6f026fbf345cac89d5fd3c69a2d6R15) does not have a great smell, and I am not sure what would be a better solution.

#### Testing
I have tried to cover the test cases that I believe are relevant. I do have two outstanding questions:
1. What testing, if any, should be done for the `ArgParser` & `HttpReqester` classes? For the `ArgParser`, I feel any test would simply be testing the ScalaJ library, and for the `ArgParser` I feel that any tests would simply be pinging the Geocoder API, which is not under the purview of unit tests.
2. What is the rule of thumb for private methods? Should they have individual tests ([scalatest supports this](https://www.scalatest.org/user_guide/using_PrivateMethodTester)) or should they be tested by virtue of testing public methods?

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -6,10 +6,17 @@ lazy val root = (project in file("."))
   .settings(
     name := "scala-challenge",
     idePackagePrefix := Some("org.hanlonjohn23"),
-    assembly / assemblyJarName := "q01.jar"
+    assembly / assemblyJarName := "q02.jar"
```
Comment (sbchapin) [inline build.sbt:9]: Note that an alternative to this would be to create [a multiproject build](https://www.scala-sbt.org/1.x/docs/Multi-Project.html) which would allow you to build independent jars and specify dependencies to each independently, allowing your old adder to exist alongside the new program.  Each jar would contain only each entrypoint and any shared code, separated by project.

File: src/main/scala/HttpRequester.scala [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+object HttpRequester {
+  case class StandardObject(
+                             staddress: JsonObject,
+                             stnumber: JsonObject,
+                             prov: String,
+                             city: String,
+                             postal: JsonObject,
+                             confidence: String
+                           )
```
Comment (sbchapin) [inline src/main/scala/HttpRequester.scala]: This contract is brittle.  You've specified `staddress`, `stnumber`,  `confidence`, and `postal`,  but done nothing with them (with the downside of making testing more difficult in addition to brittleness).  In reality, the only things you actually need are the `city` and `prov`.

The answer to making this less brittle lies in circe and its decoders.  You've relied on the semiautomatic derivation for `StandardObject`, but I would suggest writing something custom.

File: src/main/scala/HttpRequester.scala [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+  case class GeocoderResponseBody(
+                                   standard: StandardObject,
+                                   longt: String,
+                                   latt: String,
+                                   remaining_credits: String
```
Comment (sbchapin) [inline src/main/scala/HttpRequester.scala]: Similar comment here for `remaining_credits`

File: src/main/scala/HttpRequester.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  def geoCoderGetRequest(city: String): GeocoderResponseBody = {
+    val response = Http(Defaults.GEOCODER_URL)
+      .param("auth", Defaults.GEOCODER_AUTH)
+      .param("locate", city)
+      .param("json", "1")   // Request JSON response
+      .asString
+
+    implicit val standardDecoder: Decoder[StandardObject] = deriveDecoder[StandardObject]
+    implicit val gcResponseDecoder: Decoder[GeocoderResponseBody] = deriveDecoder[GeocoderResponseBody]
```
Comment (sbchapin) [inline src/main/scala/HttpRequester.scala]: Move these implicit vals out of the method.  No need to derive the decoders every time you do a request, as their definition will not change based on any parameterization - you really only need to derive them once as they depend on no parameters at all, so you may as well move em out to the scope of the HttpRequester object.

In fact, it may make sense to move them directly to companion objects for their respective classes.  But it would only make sense to do so if you put those classes in a separate module.  Telling is difficult - I'll show you why.  Grab me if/when you want to go through that.

File: src/main/scala/LeastDistanceCalculator.scala [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    def generatePairs(list: Seq[A]): Seq[(A, A)] = {
+      val head +: tail = list
+      tail.map((head, _))
+    }
+
+    list match {
+      case Nil => results
+      case _ => generateDistinctPairs(list.tail, results ++: generatePairs(list))
+    }
+  }
```
Comment (sbchapin) [inline src/main/scala/LeastDistanceCalculator.scala]: Nice usage of recursion and of type parameters here.  Dig the implementation.

---

Minor quip: I would change either the signature or the name of `generatePairs`.  It shares the same naming convention and similar signature to `generateDistinctPairs`, but its functionality and usage of head/tail is pretty different from what I would expect.

For instance, `generateDistinctPairs` will generate all distinct sets of all combinations `(a, b, c) -> (a, b), (a, c), (b, c)`, while `generatePairs` will only generate a set of the first element coupled with the rest `(a, b, c) -> (a, b), (a, c) not (b, c)`.

To distinguish the two, I'd suggest either renaming it to highlight functionality, or use a separate type signature and parameters like...
```scala
def generatePairs(left: A, rights: Seq[A]): Seq[(A, A)] = ???
```

File: src/main/scala/LeastDistanceCalculator.scala [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      // Get Distance with smallest length
+      .reduce((x,y) => getMinDistance(x, y))
+  }
+
+  private def generateDistance(locationPair: (Location, Location)): Distance = {
+    val location1 = locationPair._1
+    val location2 = locationPair._2
+
+    // Calculate Pythagorean distance
+    val distanceBetween: Double = sqrt(pow(location1.longitude - location2.longitude, 2) + pow(location1.latitude - location2.latitude, 2))
```
Comment (sbchapin) [inline src/main/scala/LeastDistanceCalculator.scala:57]: [This is how I know you're a flat-earther.](https://en.wikipedia.org/wiki/Haversine_formula)

File: src/main/scala/LeastDistanceCalculator.scala [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    generateDistinctPairs(locations)
+      // Generate Distance objects using Location pairs
+      .map(generateDistance)
+      // Get Distance with smallest length
+      .reduce((x,y) => getMinDistance(x, y))
+  }
+
+  private def generateDistance(locationPair: (Location, Location)): Distance = {
+    val location1 = locationPair._1
+    val location2 = locationPair._2
```
Comment (sbchapin) [inline src/main/scala/LeastDistanceCalculator.scala]:
```suggestion
    val (location1, location2) = locationPair
```

File: src/main/scala/LeastDistanceCalculator.scala [hunk 8]
```diff
@@ -0,0 +1,58 @@
+package org.hanlonjohn23
+
+import scala.annotation.tailrec
+import scala.math.{pow, sqrt}
+
+case class Distance(
+                     location1: Location,
+                     location2: Location,
+                     length: Double
+                   )
```
Comment (sbchapin) [inline src/main/scala/LeastDistanceCalculator.scala]: This is a very subjective comment, but has a lot to do with how you wrap data (and the relationships between domain objects):

Should a "Distance" have two locations?  Should a "Location" have distances?  Should `getClosestCities` return a Distance, or should it return Locations?

Generally, the answer to these questions can be varied and is nearly always dependent on your domain and how you need to represent it, but I generally find I am drawn to tuples if the context that binds objects together is varied and not often re-used (as I suspect this one may be).  

Specifically, `def getClosestCities(locations: Seq[Location]): (Location, Location)` satisfies the constraints of your program, and back-propagating that will eliminate some unnecessary code.

File: src/main/scala/LeastDistanceCalculator.scala [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+  private def getMinDistance(distance1: Distance, distance2: Distance): Distance = {
+    // TODO: I am sure there is a pattern that can resolve this issue in a cleaner fashion
+    val minDistanceDouble = distance1.length min distance2.length
+
+    minDistanceDouble match {
+      case distance1.length => distance1
+      case distance2.length => distance2
+    }
+  }
```
Comment (sbchapin) [inline src/main/scala/LeastDistanceCalculator.scala]: There are more complex answers to this if you are doing lensing or some other tricky operations, but I'll give you the simple one for now:  Recognize when a problem doesn't need the power of pattern matching.

```scala
  private def getMinDistance(distance1: Distance, distance2: Distance): Distance = {
    if (distance1.length < distance2.length) {
      distance1
    } else {
      distance2
    }
  }
```

File: src/test/scala/LeastDistanceCalculatorTest.scala [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+class LeastDistanceCalculatorTest extends AnyFunSuite {
+  // Fixtures
+  val losAngeles: Location = Location(city = "Los Angeles", state = "CA", latitude = 34.040462, longitude = -118.303441)
+  val oxnard: Location = Location(city = "Oxnard", state = "CA", latitude = 34.195306, longitude = -119.184146)
+  val sanDiego: Location = Location(city = "San Diego", state = "CA", latitude = 32.806472, longitude = -117.131995)
+
+  test("generateDistinctPairs should create a list of tuple distinct pairs from the passed list") {
+    assert(LeastDistanceCalculator.generateDistinctPairs(Seq("a", "b", "c"))
+      == List(("a", "b"), ("a", "c"), ("b", "c")))
```
Comment (sbchapin) [inline src/test/scala/LeastDistanceCalculatorTest.scala]: Something inadvertently being tested here is ordering.  I don't think it's a big deal, but it does mean this test could break unintentionally if your implementation changes the ordering of pairs (either within the list or within the pairs themselves).

File: src/main/scala/Location.scala [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+                     longitude: Double
+                   )
+
+object GeoResponseToLocationConverter {
+  implicit class GeocoderResponseToLocation(gcrb: GeocoderResponseBody) {
+    def toLocation: Location = Location(
+      city = gcrb.standard.city,
+      state = gcrb.standard.prov,
+      latitude = gcrb.latt.toDouble,
+      longitude = gcrb.longt.toDouble
```
Comment (sbchapin) [inline src/main/scala/Location.scala]: `toDouble` is a dangerous operation on a String.  Given that dealing with this failing is really the responsibility of the deserialization layer, I would suggest moving it there.

**Review by sbchapin — CHANGES_REQUESTED** [review]

File: src/main/scala/LeastDistanceCalculator.scala (hunk 6 above)
Comment (hanlonjohn23) [reply src/main/scala/LeastDistanceCalculator.scala:57]: I also assume there is no friction too

**Comment by sbchapin** [comment]
> ###
> #### Error Handling
> Not sure what is the best way to throw and handle errors, namely in [HttpRequestor](https://github.com/jhanlonhg/scala-challenge/compare/develop...q02#diff-d40692d2f835b6e3cb99fbc6fa0bbbce5ccc56a2e8fc4c44baa2d127be9b8e7fR42) and [Main](https://github.com/jhanlonhg/scala-challenge/compare/develop...q02#diff-a8c936a27247452dcd65a1c15bf1b96626ca6f026fbf345cac89d5fd3c69a2d6R15). Both these errors should be breaking, since they are a response error and a input error, respectively.

There are a few ways to fail, though I would argue that a response failing should not be a halting failure.

The safe way is to accept failure and provide a way around - this may be something like trying and returning `Some(thing)`, or catching, logging, and returning `None`.  Only do this if you can truly afford to fail and default, otherwise you're no better off than the same suppression pattern in java.

The "functional" way generally depends on `Try[T]` or `Either[L, R]` where [right conventionally means success](https://www.scala-lang.org/api/2.12.7/scala/util/Either.html).

The "exceptional" way generally depends on RuntimeExceptions being thrown, but the downside is that any function can fail from anywhere with a RuntimeException, which is generally a bad thing.

Generally, I tend to push failure to places that should be able to fail, and keep that failure there and test it.  In your case, I imagine an `HttpRequestor` encapsulating that failure in its response rather than escalating it.  I also imagine moving the `IllegalArgumentException` from Main up into to the arg parser where it is actually failing (from a responsibility perspective) - enforce that the arguments have `cities: Seq[String]` and do the parsing immediately in the arg parser.

> #### Bad Patterns
> [This solution](https://github.com/jhanlonhg/scala-challenge/compare/develop...q02#diff-a8c936a27247452dcd65a1c15bf1b96626ca6f026fbf345cac89d5fd3c69a2d6R15) does not have a great smell, and I am not sure what would be a better solution.

I didn't want to go crazy with the answer [here](https://github.com/jhanlonhg/scala-challenge/pull/2#discussion_r804246954), but if you want a further explanation, please ask.

> #### Testing
> I have tried to cover the test cases that I believe are relevant. I do have two outstanding questions:
> 
> 1. What testing, if any, should be done for the `ArgParser` & `HttpReqester` classes? For the `ArgParser`, I feel any test would simply be testing the ScalaJ library, and for the `ArgParser` I feel that any tests would simply be pinging the Geocoder API, which is not under the purview of unit tests.
> 2. What is the rule of thumb for private methods? Should they have individual tests ([scalatest supports this](https://www.scalatest.org/user_guide/using_PrivateMethodTester)) or should they be tested by virtue of testing public methods?

1. I would test the "good paths" of `ArgParser` behaviorally.  You want the Arguments instance to contain formatted, parsed values.  That way you can swap your arg parser library out whenever you want, since none of its implementation details are really leaking.  `HttpRequester` can [be mocked](https://www.scalatest.org/user_guide/testing_with_mock_objects), but I personally find a rather marginal amount of value in that.
2. Don't test private methods.  At worst, test protected or package-local methods, but avoid it.  If something is sufficiently complicated enough that testing many of the edge cases within private methods is difficult, consider splitting your solution into smaller pieces, each of which can be tested (then, behaviorally test them all together re-assembled).

File: src/main/scala/LeastDistanceCalculator.scala (hunk 6 above)
Comment (sbchapin) [reply src/main/scala/LeastDistanceCalculator.scala:57]: Also a problem is negative/positive lat/lon.

Here's a good testcase for you: 

 What is closer?
- [Adak, Alaska United States](https://www.google.com/maps/search/?api=1&query=Adak%2C%20Alaska%20United%20States)Coordinates: 51°53′N 176°39′W
- [Anadyr, Chukotka Autonomous Okrug Russia](https://www.google.com/maps/search/?api=1&query=Anadyr%2C%20Chukotka%20Autonomous%20Okrug%20Russia), Coordinates: 64°44′N 177°31′E
- Los Angeles?

Basically, the edge-cases where lat/long wrap around.  [Use this](https://blog.batchgeo.com/cities-by-latitude-and-longitude/) to find out.

File: src/main/scala/HttpRequester.scala (hunk 2 above)
Comment (hanlonjohn23) [reply src/main/scala/HttpRequester.scala]: This structure has replaced with the Location object, which only contains relevant information.

File: src/main/scala/HttpRequester.scala (hunk 3 above)
Comment (hanlonjohn23) [reply src/main/scala/HttpRequester.scala]: This structure has replaced with the Location object, which only contains relevant information.

File: src/main/scala/LeastDistanceCalculator.scala (hunk 5 above)
Comment (hanlonjohn23) [reply src/main/scala/LeastDistanceCalculator.scala]: Changed the name and the signature!

File: src/main/scala/LeastDistanceCalculator.scala (hunk 8 above)
Comment (hanlonjohn23) [reply src/main/scala/LeastDistanceCalculator.scala]: Made Distance a private member

File: src/main/scala/LeastDistanceCalculator.scala (hunk 9 above)
Comment (hanlonjohn23) [reply src/main/scala/LeastDistanceCalculator.scala]: 👍

File: src/main/scala/HttpRequester.scala (hunk 4 above)
Comment (hanlonjohn23) [reply src/main/scala/HttpRequester.scala]: Resolved by using a custom decoder, and calling that decoder in a higher scope.

File: src/main/scala/LeastDistanceCalculator.scala (hunk 6 above)
Comment (hanlonjohn23) [reply src/main/scala/LeastDistanceCalculator.scala:57]: Updated code to use Haversine Distance calculations, and added a test for distances over the 180th meridian.

File: src/test/scala/LeastDistanceCalculatorTest.scala [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+  test("getClosestCities should return a tuple containing the closest Location objects") {
+    assert(LeastDistanceCalculator.getClosestCities(Seq(losAngeles, sanDiego, oxnard))
+      == (losAngeles, oxnard) )
+  }
+
+  test("getClosestCities should handle cities on opposite sides of the 180th meridian") {
+    assert(LeastDistanceCalculator.getClosestCities(Seq(honolulu, tokyo, newYork))
+    == (honolulu, tokyo) )
+  }
```
Comment (sbchapin) [inline src/test/scala/LeastDistanceCalculatorTest.scala:28]: In love with these tests

File: src/test/scala/LocationTest.scala [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+        |	},
+        |	"longt": "-118.303441",
+        |	"latt": "34.040462"
+        |}
+        |""".stripMargin
+
+    val location = Location(
+      city = "Los Angeles", state = "CA", latitude = 34.040462, longitude = -118.303441
+    )
+    assert(decode[Location](geocoderResponseBody).right.getOrElse() == location)
```
Comment (sbchapin) [inline src/test/scala/LocationTest.scala]: Alternatively, 
```scala
assert(decode[Location](geocoderResponseBody) == Right(location))
```

Then it will tell you what things both didn't equal rather than just exception barfing on getOrElse()

**Review by sbchapin — APPROVED** [review]
Solid.

File: src/test/scala/LocationTest.scala (hunk 13 above)
Comment (hanlonjohn23) [reply src/test/scala/LocationTest.scala]: Great advice. Implemented.

File: src/test/scala/LeastDistanceCalculatorTest.scala (hunk 12 above)
Comment (hanlonjohn23) [reply src/test/scala/LeastDistanceCalculatorTest.scala:28]: Pretty proud myself

File: src/main/scala/Location.scala (hunk 11 above)
Comment (hanlonjohn23) [reply src/main/scala/Location.scala]: Resolved as advised.
